### PR TITLE
Allow AVLJs to perform judge tasks on legacy appeals

### LINKED ISSUE #11971: Allow AVLJs to be assigned legacy cases
Started from [this slack thread](https://dsva.slack.com/archives/CHX8FMP28/p1567534420110500). 

For attorneys and judges, we determine what type of task (and therefor what actions are available to them) for a legacy appeal assigned to them [based on their vacols role](https://github.com/department-of-veterans-affairs/caseflow/blob/a86c1e04e52faf6911b99281908576790fc692df/app/models/queues/legacy_work_queue.rb#L30). This causes an issue for acting judges as their first role is always "attorney" and they do not receive the ability to dispatch a drafted appeal.

This ticket is to determine and implement how we decide which actions an acting judge sees.

### Expand conditional to include judge user

By using `user&.vacols_roles&.include? "judge"  we would select all judges and acting judges. However, this would display a judge task and actions to every acting judge that is assigned a legacy task, regardless if it was assigned intending to be a judge task or an attorney task (essentially the opposite of what happens now).

### Using JudgeTeam ownership to determine acting judges

We do this in [various other places](https://github.com/department-of-veterans-affairs/caseflow/search?q=JudgeTeam.for_judge&unscoped_q=JudgeTeam.for_judge) in the code where we want to allow AVLJs to have judge functionality. This would require us to keep the JudgeTeams for acting judges up to date, but would allow us to have some control over who is currently working on drafting a decision vs evaluating one.

### Using issues with decided dispositions

An appeal where all issues have non nil dispositions for their issues is guaranteed to have gone through attorney checkout. This would cause some weirdness when adding issues after some of them have had dispositions added.

### Using appeals with decision documents

An appeal with a decision document is guaranteed to have gone through attorney checkout. This would break as soon as a judge sends the case back to the attorney.

## Actually creating the tasks and assigning them to users as we do with ama appeals

Not exactly a viable solution for the AVLJs working AOJ cases who needed a solution to this
around yesterday.

## Thoughts

This was mostly to get some things from my mind before stepping away for the day. Will revisit in the morning for feedback and ideas from engineering, support, and product.

### PR DESCRIPTION
**Pull request by hschallhorn**
Resolves #11971

### Description
Due to the way we determine [task type for legacy work queues](https://github.com/department-of-veterans-affairs/caseflow/blob/a86c1e04e52faf6911b99281908576790fc692df/app/models/queues/legacy_work_queue.rb#L30), acting judges have been receiving judge tasks, but cannot act upon them. This PR classes any appeal with an associated AttorneyCaseReview assigned to someone with a judge vacols_role as a JudgeTask, allowing AVLJs to review legacy cases.

### CODE REVIEW

File: app/models/queues/legacy_work_queue.rb [hunk 1]
```diff
@@ -40,5 +40,13 @@ def validate_or_create_user(user, css_id)
         User.find_by_css_id_or_create_with_default_station_id(css_id)
       end
     end
+
+    def issues_have_dispositions(appeal)
+      appeal.issues.all? { |issue| !issue.disposition.nil? }
+    end
```
Comment (hschallhorn) [inline app/models/queues/legacy_work_queue.rb]: If all of the issues for an appeal have a disposition, the decision has already been drafted by an attorney.  We are then assuming that this task has been assigned to an AVLJ for the purpose of them performing a judge task.

Is there a better metric to use for this?

File: app/controllers/legacy_tasks_controller.rb [hunk 2]
```diff
@@ -98,7 +98,7 @@ def json_task(task)
     ::WorkQueue::LegacyTaskSerializer.new(task)
   end
 
-  def json_tasks(tasks, role)
-    ::WorkQueue::LegacyTaskSerializer.new(tasks, is_collection: true, params: { role: role })
+  def json_tasks(tasks, user, role)
+    ::WorkQueue::LegacyTaskSerializer.new(tasks, is_collection: true, params: { user: user, role: role })
```
Comment (hschallhorn) [inline app/controllers/legacy_tasks_controller.rb]: Ensures the current user gets passed to `task.available_actions`

File: app/models/queues/legacy_work_queue.rb [hunk 3]
```diff
@@ -27,7 +27,7 @@ def tasks_from_vacols_tasks(vacols_tasks, user = nil)
       vacols_tasks.zip(vacols_appeals).map do |task, appeal|
         user = validate_or_create_user(user, task.assigned_to_css_id)
 
-        task_class = (user&.vacols_roles&.first&.downcase == "judge") ? JudgeLegacyTask : AttorneyLegacyTask
+        task_class = (user_is_judge(user) && issues_have_dispositions(appeal)) ? JudgeLegacyTask : AttorneyLegacyTask
```
Comment (leikkisa) [inline app/models/queues/legacy_work_queue.rb]: Based on your comment below, should this be an || ?

File: app/models/queues/legacy_work_queue.rb (hunk 1 above)
Comment (leikkisa) [reply app/models/queues/legacy_work_queue.rb]: I think `issue.disposition.present?` would be better than the double negative.

File: app/models/queues/legacy_work_queue.rb [hunk 4]
```diff
@@ -40,5 +40,13 @@ def validate_or_create_user(user, css_id)
         User.find_by_css_id_or_create_with_default_station_id(css_id)
       end
     end
+
+    def issues_have_dispositions(appeal)
+      appeal.issues.all? { |issue| !issue.disposition.nil? }
+    end
+
+    def user_is_judge(user)
+      user&.vacols_roles&.any? { |role| role.casecmp("judge").zero? }
```
Comment (leikkisa) [inline app/models/queues/legacy_work_queue.rb]: I like that you're not just checking the first one now! Good bug prevention

File: app/models/queues/legacy_work_queue.rb (hunk 3 above)
Comment (hschallhorn) [reply app/models/queues/legacy_work_queue.rb]: So we're actually changing the way we classify a user as a judge here. Previously we were classifying all acting judges (who will have the vacols_roles ["attorney", "judge"]) as just attorneys. So here we're now checking that the user has a judge role.

But, as acting judges still can receive attorney tasks, we want to ensure we aren't incorrectly giving them judge tasks for decisions that haven't yet been drafted, so we want to make sure both of these conditionals are true.

File: app/models/legacy_tasks/judge_legacy_decision_review_task.rb [hunk 5]
```diff
@@ -9,8 +9,8 @@ def review_action
     end
   end
 
-  def available_actions(current_user, role)
-    return [] if role != "judge" || current_user != assigned_to
+  def available_actions(current_user, _role)
+    return [] if current_user.vacols_roles.none? { |vrole| vrole.casecmp("judge").zero? } || current_user != assigned_to
```
Comment (pkarman) [inline app/models/legacy_tasks/judge_legacy_decision_review_task.rb]: I would love to see this logic on the `User` model instead, encapsulated in a well-named method.

maybe something like `current_user.acts_like_vacols_judge?` or something?

File: app/models/queues/legacy_work_queue.rb (hunk 1 above)
Comment (pkarman) [reply app/models/queues/legacy_work_queue.rb]: this method seems like it belongs on the `appeal` class, not as a helper in this one.

File: app/models/queues/legacy_work_queue.rb (hunk 4 above)
Comment (pkarman) [reply app/models/queues/legacy_work_queue.rb]: again, this method could go on the `User` model.

File: app/models/legacy_tasks/judge_legacy_decision_review_task.rb (hunk 5 above)
Comment (hschallhorn) [reply app/models/legacy_tasks/judge_legacy_decision_review_task.rb]: 🤦‍♀  or the already existant `judge_in_vacols?`

File: app/models/legacy_appeal.rb [hunk 6]
```diff
@@ -733,6 +733,18 @@ def paper_case?
...
+  end
+
+  def vacols_case_review_creation_date_in_string_format
+    VacolsHelper.day_only_str(vacols_case_review.created_at)
+  end
+
+  def vacols_case_review
+    VACOLS::CaseAssignment.latest_task_for_appeal(vacols_id)
+  end
+
```
Comment (hschallhorn) [inline app/models/legacy_appeal.rb]: Moved relevant case review stuff out of update_legacy_attorney_case_review to here

File: app/models/queues/legacy_work_queue.rb [hunk 7]
```diff
@@ -27,7 +27,7 @@ def tasks_from_vacols_tasks(vacols_tasks, user = nil)
       vacols_tasks.zip(vacols_appeals).map do |task, appeal|
         user = validate_or_create_user(user, task.assigned_to_css_id)
 
-        task_class = (user&.vacols_roles&.first&.downcase == "judge") ? JudgeLegacyTask : AttorneyLegacyTask
+        task_class = (user.judge_in_vacols? && appeal.attorney_case_review) ? JudgeLegacyTask : AttorneyLegacyTask
```
Comment (hschallhorn) [inline app/models/queues/legacy_work_queue.rb]: Crux of the change. If a legacy case is assigned to an avlj in vacols and the case has already been reviewed by an attorney, we will consider this task to be a JudgeLegacyTask.

File: app/workflows/update_legacy_attorney_case_review.rb [hunk 8]
```diff
@@ -30,38 +30,26 @@ def call
...
   def vacols_case_review_exists
     return if vacols_case_review
 
     errors.add(:document_id, "Could not find a legacy Attorney Case Review with id #{id}")
   end
 
   def vacols_case_review
-    @vacols_case_review ||= VACOLS::CaseAssignment.latest_task_for_appeal(id)
+    @vacols_case_review ||= appeal.vacols_case_review
   end
```
Comment (hschallhorn) [inline app/workflows/update_legacy_attorney_case_review.rb]: Moved all this to legacy_appeal

**Pull request merged**